logr.go

```diff
@@ -371,12 +383,63 @@ func (l Logger) IsZero() bool {
 	return l.sink == nil
 }
 
+// WithContext stores a context in the Logger. If the Logger also has keys set
+// that it is meant to log from a context, then the values for those keys from
+// the given context will be added to all log entries.
+func (l Logger) WithContext(ctx context.Context) Logger {
+	l.ctx = ctx
+	return l
+}
+
+// WithContextValues extends the list of context keys and the name for them. When
+// given a context through WithContext later, the Logger will extract the
+// values for these keys and log them as additional key/value pairs.
+func (l Logger) WithContextValues(keys ...ContextKey) Logger {
+	if l.contextKeys == nil {
+		// Copy the parameters to avoid surprises when the caller changes the content
+		// of the parameter slice later.
+		contextKeys := make([]ContextKey, 0, len(keys))
+		contextKeys = append(contextKeys, keys...)
+		l.contextKeys = &contextKeys
+	} else {
+		// We must create a new slice because the existing one is shared between
+		// Logger instances.
+		contextKeys := make([]ContextKey, 0, len(*l.contextKeys)+len(keys))
+		contextKeys = append(contextKeys, *l.contextKeys...)
+		contextKeys = append(contextKeys, keys...)
+		l.contextKeys = &contextKeys
+	}
+	return l
+}
+
+func (l Logger) valuesFromContext() []interface{} {
+	if l.ctx == nil || l.contextKeys == nil {
+		return nil
+	}
+	keysAndValues := make([]interface{}, 0, 2*len(*l.contextKeys))
+	for _, key := range *l.contextKeys {
+		if value := l.ctx.Value(key.Key); value != nil {
+			keysAndValues = append(keysAndValues, key.Name, value)
+		}
+	}
+	return keysAndValues
+}
+
+// ContextKey defines the string that is to be used in a key/value pair when
+// logging the value that is stored in a context for the given key.
+type ContextKey struct {
```

Comment:
We could define `WithContextValues` as taking `keysAndNames ...interface{}` and then wouldn't need this type. But I thought that this would be much too easy to confuse with `keysAndValues ...interface{}` and therefore decided to use a type-safe API.